src/libstd/sys/unix/fs.rs

```diff
@@ -132,6 +139,32 @@ impl FromInner<raw::mode_t> for FilePermissions {
 impl Iterator for ReadDir {
     type Item = io::Result<DirEntry>;
 
+    #[cfg(target_os = "sunos")]
```

Comment:
@alexcrichton, as it seems, the implementation can't be shared - not because of `readdir_r` (it does have a standard-compliant version), but because of a quite interesting `dirent` structure. This struct def on Solaris/Illumos contains a zero-length array:

    typedef struct dirent {
      ino_t           d_ino;          /* "inode number" of entry */
      off_t           d_off;          /* offset of disk directory entry */
      unsigned short  d_reclen;       /* length of this record */
      char            d_name[1];      /* name of file */
    } dirent_t;

Notice the `d_name` field of length 1. I asked a [question on this issue](https://users.rust-lang.org/t/zero-length-arrays-rust-ffi/4051) on the Rust users forum, but haven't found a definite answer. Experiments showed that Rust supports such structures, but not when they're nested in another structure - as is the case with `std::fs::DirEntry` and `libc::dirent`.

So, I guess, a separate implementation would be required, and as it's OS-specific anyway, I used the Solaris/Illumos-specific `readdir` function, as recommended by the man page.

What do you think?